File: android/app/src/main/java/net/bradball/teetimecaddie/android/ui/common/buttons/TtcOutlinedButton.kt
package net.bradball.teetimecaddie.android.ui.common.buttons

import android.content.res.Configuration
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import net.bradball.teetimecaddie.android.theme.TtcColorRole
import net.bradball.teetimecaddie.android.theme.MyApplicationTheme
import net.bradball.teetimecaddie.android.ui.common.LoadingIndicator
import net.bradball.teetimecaddie.android.ui.common.icons.TtcIcons

/**
 * An outlined pill button from the Fairway Morning design system.
 *
 * By default ([TtcColorRole.Neutral]) it renders the design's neutral treatment: on-surface text with
 * an `outline` border. Every other role tints both the text and the border to that role — including
 * [TtcColorRole.Error], for a low-emphasis destructive action. Use [icon] for an optional leading icon, [dense] for tighter
 * padding, and [isLoading] to show an animated indicator in place of the label (clicks are ignored while
 * loading). For a full-width CTA, pass `Modifier.fillMaxWidth()`.
 *
 * @param text The button label.
 * @param onClick Invoked when the button is tapped (ignored while [isLoading]).
 * @param modifier Modifier for the button.
 * @param color The [TtcColorRole] role. Defaults to [TtcColorRole.Neutral].
 * @param icon Optional leading [TtcIcons] icon.
 * @param dense When true, uses tighter content padding.
 * @param enabled Whether the button is enabled.
 * @param isLoading When true, shows a loading indicator in place of the label and ignores clicks.
 */
@Composable
fun TtcOutlinedButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    color: TtcColorRole = TtcColorRole.Neutral,
    icon: TtcIcons? = null,
    dense: Boolean = false,
    enabled: Boolean = true,
    isLoading: Boolean = false,
) {
    OutlinedButton(
        onClick = { if (!isLoading) onClick() },
        modifier = modifier,
        enabled = enabled,
        shape = TtcButtonDefaults.Shape,
        colors = TtcButtonDefaults.outlinedColors(color),
        border = TtcButtonDefaults.outlinedBorder(color, enabled),
        contentPadding = TtcButtonDefaults.contentPadding(dense),
    ) {
        LoadingIndicator(isLoading = isLoading) {
            if (icon != null) {
                Icon(
                    painter = icon.painter,
                    contentDescription = null,
                    modifier = Modifier.size(TtcButtonDefaults.IconSize),
                )
                Spacer(Modifier.width(TtcButtonDefaults.IconSpacing))
            }
            Text(text)
        }
    }
}

@Preview(name = "Light", showBackground = true)
@Preview(name = "Dark", showBackground = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun TtcOutlinedButtonPreview() {
    MyApplicationTheme {
        Surface {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                TtcOutlinedButton(text = "Invite more", onClick = {})
                TtcOutlinedButton(text = "Invite", onClick = {}, icon = TtcIcons.ADD)
                TtcOutlinedButton(text = "Primary", onClick = {}, color = TtcColorRole.Primary)
                TtcOutlinedButton(text = "Leave game", onClick = {}, color = TtcColorRole.Error)
                TtcOutlinedButton(text = "Create account", onClick = {}, modifier = Modifier.fillMaxWidth())
            }
        }
    }
}

@Preview(name = "States - Light", showBackground = true)
@Preview(name = "States - Dark", showBackground = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun TtcOutlinedButtonStatesPreview() {
    MyApplicationTheme {
        Surface {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                TtcOutlinedButton(text = "Enabled", onClick = {})
                TtcOutlinedButton(text = "Disabled", onClick = {}, enabled = false)
                TtcOutlinedButton(text = "Loading", onClick = {}, isLoading = true)
            }
        }
    }
}

File: ios/TeeTimeCaddie/TeeTimeCaddie/Source/ui/common/Buttons/TtcOutlinedButton.swift
import SwiftUI
import ThemeUI

/// An outlined pill button from the Fairway Morning design system.
///
/// By default (`.neutral`) it renders the design's neutral treatment: on-surface text with a subtle
/// `outline` border. Every other ``TtcColorRole`` tints both the text and the border to that role —
/// including `.error`, for a low-emphasis destructive action.
/// Use [icon] for an optional leading icon, [dense] for tighter padding, and [isLoading] to show an
/// animated indicator in place of the label (taps are ignored while loading). For a full-width CTA,
/// apply `.frame(maxWidth: .infinity)`.
struct TtcOutlinedButton: View {
    @EnvironmentObject private var theme: AppTheme

    private let title: String
    private let color: TtcColorRole
    private let icon: ImageResource?
    private let dense: Bool
    private let isLoading: Bool
    private let action: () -> Void

    init(
        _ title: String,
        color: TtcColorRole = .neutral,
        icon: ImageResource? = nil,
        dense: Bool = false,
        isLoading: Bool = false,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.color = color
        self.icon = icon
        self.dense = dense
        self.isLoading = isLoading
        self.action = action
    }

    var body: some View {
        let scheme = theme.colorScheme
        Button(action: { if !isLoading { action() } }) {
            TtcButtonLabel(title: title, icon: icon)
                .loadingOverlay(type: .Flashing, isLoading: isLoading)
        }
        .buttonStyle(TtcPillButtonStyle(
            background: .clear,
            foreground: TtcButtonStyle.outlinedForeground(color, scheme),
            disabledBackground: .clear,
            disabledForeground: scheme.onSurfaceVariant,
            border: TtcButtonStyle.outlinedBorder(color, scheme),
            disabledBorder: scheme.outlineVariant,
            dense: dense
        ))
    }
}

#Preview("Light") {
    TeeTimeCaddieTheme {
        VStack(alignment: .leading, spacing: 12) {
            TtcOutlinedButton("Invite more") {}
            TtcOutlinedButton("Invite", icon: .Icons.calendarAdd) {}
            TtcOutlinedButton("Primary", color: .primary) {}
            TtcOutlinedButton("Leave game", color: .error) {}
            TtcOutlinedButton("Create account") {}.frame(maxWidth: .infinity)
            TtcOutlinedButton("Disabled") {}.disabled(true)
            TtcOutlinedButton("Loading", isLoading: true) {}
        }
        .padding()
    }
}

#Preview("Dark") {
    TeeTimeCaddieTheme {
        VStack(alignment: .leading, spacing: 12) {
            TtcOutlinedButton("Invite more") {}
            TtcOutlinedButton("Primary", color: .primary) {}
            TtcOutlinedButton("Disabled") {}.disabled(true)
            TtcOutlinedButton("Loading", isLoading: true) {}
        }
        .padding()
    }
    .preferredColorScheme(.dark)
}
